File: src/lib/comparisonEngine.ts
import {
  ComparisonResult,
  ComparisonRow,
  CellChange,
  ParsedFile,
} from "./types";

function stringify(value: unknown): string {
  if (value === null || value === undefined) return "";
  return String(value);
}

export function compareFiles(
  before: ParsedFile,
  after: ParsedFile,
  keyColumn: string
): ComparisonResult {
  const allColumns = Array.from(
    new Set([...before.headers, ...after.headers])
  );

  // Build maps keyed by the key column
  const beforeMap = new Map<string, Record<string, unknown>>();
  const afterMap = new Map<string, Record<string, unknown>>();

  for (const row of before.data) {
    const key = stringify(row[keyColumn]);
    if (key) beforeMap.set(key, row);
  }

  for (const row of after.data) {
    const key = stringify(row[keyColumn]);
    if (key) afterMap.set(key, row);
  }

  const allKeys = Array.from(
    new Set([...beforeMap.keys(), ...afterMap.keys()])
  );

  const rows: ComparisonRow[] = [];
  let added = 0;
  let deleted = 0;
  let modified = 0;
  let unchanged = 0;

  for (let i = 0; i < allKeys.length; i++) {
    const key = allKeys[i];
    const beforeRow = beforeMap.get(key) || null;
    const afterRow = afterMap.get(key) || null;

    if (!beforeRow && afterRow) {
      // Added
      rows.push({
        rowIndex: i,
        status: "added",
        keyValue: key,
        beforeData: null,
        afterData: afterRow,
        changes: [],
      });
      added++;
    } else if (beforeRow && !afterRow) {
      // Deleted
      rows.push({
        rowIndex: i,
        status: "deleted",
        keyValue: key,
        beforeData: beforeRow,
        afterData: null,
        changes: [],
      });
      deleted++;
    } else if (beforeRow && afterRow) {
      // Check for modifications
      const changes: CellChange[] = [];
      for (const col of allColumns) {
        const oldVal = stringify(beforeRow[col]);
        const newVal = stringify(afterRow[col]);
        if (oldVal !== newVal) {
          changes.push({
            column: col,
            oldValue: beforeRow[col] as string | number | boolean | null,
            newValue: afterRow[col] as string | number | boolean | null,
          });
        }
      }

      if (changes.length > 0) {
        rows.push({
          rowIndex: i,
          status: "modified",
          keyValue: key,
          beforeData: beforeRow,
          afterData: afterRow,
          changes,
        });
        modified++;
      } else {
        rows.push({
          rowIndex: i,
          status: "unchanged",
          keyValue: key,
          beforeData: beforeRow,
          afterData: afterRow,
          changes: [],
        });
        unchanged++;
      }
    }
  }

  // Sort: deleted first, modified, added, unchanged
  const statusOrder = { deleted: 0, modified: 1, added: 2, unchanged: 3 };
  rows.sort((a, b) => statusOrder[a.status] - statusOrder[b.status]);

  return {
    columns: allColumns,
    keyColumn,
    rows,
    summary: {
      total: allKeys.length,
      added,
      deleted,
      modified,
      unchanged,
    },
    beforeFileName: before.fileName,
    afterFileName: after.fileName,
    comparedAt: new Date().toISOString(),
  };
}

File: comparison_engine.py
"""Core comparison engine for detecting changes between two datasets."""

from datetime import datetime


def stringify(value):
    """Convert any value to string for comparison."""
    if value is None or (isinstance(value, float) and str(value) == "nan"):
        return ""
    return str(value)


def compare_datasets(before_data, after_data, key_column, columns):
    """
    Compare two datasets row-by-row using a key column.

    Returns a dict with:
      - columns: list of all column names
      - keyColumn: the key column used
      - rows: list of comparison row dicts
      - summary: counts of added/deleted/modified/unchanged
    """
    # Build lookup maps
    before_map = {}
    for row in before_data:
        key = stringify(row.get(key_column, ""))
        if key:
            before_map[key] = row

    after_map = {}
    for row in after_data:
        key = stringify(row.get(key_column, ""))
        if key:
            after_map[key] = row

    all_keys = list(dict.fromkeys(list(before_map.keys()) + list(after_map.keys())))

    rows = []
    added = 0
    deleted = 0
    modified = 0
    unchanged = 0

    for i, key in enumerate(all_keys):
        before_row = before_map.get(key)
        after_row = after_map.get(key)

        if before_row is None and after_row is not None:
            rows.append({
                "rowIndex": i,
                "status": "added",
                "keyValue": key,
                "beforeData": None,
                "afterData": after_row,
                "changes": [],
            })
            added += 1

        elif before_row is not None and after_row is None:
            rows.append({
                "rowIndex": i,
                "status": "deleted",
                "keyValue": key,
                "beforeData": before_row,
                "afterData": None,
                "changes": [],
            })
            deleted += 1

        elif before_row is not None and after_row is not None:
            changes = []
            for col in columns:
                old_val = stringify(before_row.get(col, ""))
                new_val = stringify(after_row.get(col, ""))
                if old_val != new_val:
                    changes.append({
                        "column": col,
                        "oldValue": before_row.get(col, ""),
                        "newValue": after_row.get(col, ""),
                    })

            if changes:
                rows.append({
                    "rowIndex": i,
                    "status": "modified",
                    "keyValue": key,
                    "beforeData": before_row,
                    "afterData": after_row,
                    "changes": changes,
                })
                modified += 1
            else:
                rows.append({
                    "rowIndex": i,
                    "status": "unchanged",
                    "keyValue": key,
                    "beforeData": before_row,
                    "afterData": after_row,
                    "changes": [],
                })
                unchanged += 1

    # Sort: deleted, modified, added, unchanged
    status_order = {"deleted": 0, "modified": 1, "added": 2, "unchanged": 3}
    rows.sort(key=lambda r: status_order.get(r["status"], 4))

    return {
        "columns": columns,
        "keyColumn": key_column,
        "rows": rows,
        "summary": {
            "total": len(all_keys),
            "added": added,
            "deleted": deleted,
            "modified": modified,
            "unchanged": unchanged,
        },
        "comparedAt": datetime.now().isoformat(),
    }
